sled-storage/src/manager.rs

```diff
@@ -764,6 +973,77 @@ impl StorageManager {
         }
     }
 
+    // Ensures a dataset exists within a zpool, according to `config`.
+    async fn ensure_dataset(
+        &mut self,
+        config: &DatasetConfig,
+    ) -> Result<(), Error> {
+        info!(self.log, "ensure_dataset"; "config" => ?config);
+
+        // We can only place datasets within managed disks.
+        // If a disk is attached to this sled, but not a part of the Control
+        // Plane, it is treated as "not found" for dataset placement.
+        if !self
+            .resources
+            .disks()
+            .iter_managed()
+            .any(|(_, disk)| disk.zpool_name() == config.name.pool())
+        {
+            return Err(Error::ZpoolNotFound(format!(
+                "{}",
+                config.name.pool(),
+            )));
+        }
+
+        let zoned = config.name.dataset().zoned();
+        let mountpoint_path = if zoned {
+            Utf8PathBuf::from("/data")
+        } else {
+            config.name.pool().dataset_mountpoint(
+                &Utf8PathBuf::from("/"),
+                &config.name.dataset().to_string(),
+            )
+        };
+        let mountpoint = Mountpoint::Path(mountpoint_path);
+
+        let fs_name = &config.name.full_name();
+        let do_format = true;
+
+        // The "crypt" dataset needs these details, but should already exist
+        // by the time we're creating datasets inside.
+        let encryption_details = None;
```

Comment:
So `None` here does **not** mean "remove encryption"? Is there a way or a need to do that?

Reply:
The `Zfs::ensure_filesystem` function probably needs some re-work, I agree that this is confusing...

No, this isn't necessary, at least not for this PR. We only supply a value here for "encryption roots" (e.g., the `crypt` root) which is still automatically created by sled agents.

All our encrypted filesystems used dataset names within the `crypt/` dataset, and are implicitly encrypted.

See: `DatasetName::full_name`, which puts the dataset on `crypt/` if it **should** be encrypted.

I'll add some extra docs to `ensure_filesystem`, in the meanwhile.